File: src/DevProjects.Core/Services/EnvFileEditor.cs
namespace DevProjects.Core.Services;

/// <summary>
/// Reads and edits a .env file as KEY=VALUE lines, preserving comments, blank lines, and order.
/// Pure string operations over the file text — the UI handles read/write. Values are returned
/// verbatim (no unquoting) to avoid corrupting secrets on round-trip.
/// </summary>
public static class EnvFileEditor
{
    public readonly record struct EnvEntry(string Key, string Value);

    public static IReadOnlyList<EnvEntry> Parse(string text)
    {
        var result = new List<EnvEntry>();
        foreach (var raw in Normalize(text).Split('\n'))
        {
            var line = raw.TrimStart();
            if (line.Length == 0 || line.StartsWith('#')) continue;
            var eq = line.IndexOf('=');
            if (eq <= 0) continue;
            result.Add(new EnvEntry(line[..eq].Trim(), line[(eq + 1)..]));
        }
        return result;
    }

    public static string SetKey(string text, string key, string value)
    {
        var lines = SplitKeepingTrailing(Normalize(text));
        for (var i = 0; i < lines.Count; i++)
        {
            if (IsAssignmentFor(lines[i], key)) { lines[i] = $"{key}={value}"; return string.Join('\n', lines); }
        }
        // Append before any trailing empty element so we keep a single final newline.
        var insertAt = lines.Count > 0 && lines[^1].Length == 0 ? lines.Count - 1 : lines.Count;
        lines.Insert(insertAt, $"{key}={value}");
        return string.Join('\n', lines);
    }

    public static string RemoveKey(string text, string key)
    {
        var lines = SplitKeepingTrailing(Normalize(text));
        lines.RemoveAll(l => IsAssignmentFor(l, key));
        return string.Join('\n', lines);
    }

    private static bool IsAssignmentFor(string line, string key)
    {
        var t = line.TrimStart();
        var eq = t.IndexOf('=');
        return eq > 0 && string.Equals(t[..eq].Trim(), key, StringComparison.Ordinal);
    }

    private static string Normalize(string text) => text.Replace("\r\n", "\n");

    // Preserve a trailing newline as a final empty element so round-trips don't strip it.
    private static List<string> SplitKeepingTrailing(string text) => text.Split('\n').ToList();
}

File: electron/src/core/config/envFileEditor.ts
/**
 * Reads and edits .env file content as KEY=VALUE lines, preserving comments,
 * blank lines, and order. Pure string operations — file I/O is the caller's
 * responsibility. Values are returned verbatim (no unquoting) to avoid
 * corrupting secrets on round-trip.
 *
 * Direct port of C# EnvFileEditor.
 */

/** One parsed KEY=VALUE pair from a .env file. */
export interface EnvEntry {
  key: string
  value: string
}

function normalize(text: string): string {
  return text.replace(/\r\n/g, '\n')
}

/** Split on '\n' preserving the trailing empty element (keeps final newline on round-trip). */
function splitKeepingTrailing(text: string): string[] {
  return text.split('\n')
}

function isAssignmentFor(line: string, key: string): boolean {
  const t = line.trimStart()
  const eq = t.indexOf('=')
  if (eq <= 0) return false
  return t.slice(0, eq).trim() === key  // ordinal / case-sensitive, same as C#
}

/**
 * Parses KEY=VALUE lines from .env text. Skips comments (#) and blank lines.
 * Values are returned verbatim (no unquoting).
 */
export function parse(text: string): EnvEntry[] {
  const result: EnvEntry[] = []
  for (const raw of normalize(text).split('\n')) {
    const line = raw.trimStart()
    if (line.length === 0 || line.startsWith('#')) continue
    const eq = line.indexOf('=')
    if (eq <= 0) continue
    result.push({ key: line.slice(0, eq).trim(), value: line.slice(eq + 1) })
  }
  return result
}

/**
 * Replaces the value of an existing key in-place, or appends a new KEY=VALUE
 * line. Preserves comments, blank lines, and ordering. Preserves trailing newline.
 */
export function setKey(text: string, key: string, value: string): string {
  const lines = splitKeepingTrailing(normalize(text))
  for (let i = 0; i < lines.length; i++) {
    if (isAssignmentFor(lines[i], key)) {
      lines[i] = `${key}=${value}`
      return lines.join('\n')
    }
  }
  // Insert before any trailing empty element so we keep a single final newline.
  const insertAt =
    lines.length > 0 && lines[lines.length - 1].length === 0
      ? lines.length - 1
      : lines.length
  lines.splice(insertAt, 0, `${key}=${value}`)
  return lines.join('\n')
}

/**
 * Removes all lines matching the given key. Preserves everything else.
 */
export function removeKey(text: string, key: string): string {
  const lines = splitKeepingTrailing(normalize(text))
  return lines.filter(l => !isAssignmentFor(l, key)).join('\n')
}
